src/System Application/App/MicrosoftGraph/src/GraphOptionalParameters.Codeunit.al

```diff
@@ -72,23 +69,120 @@ codeunit 9353 "Graph Optional Parameters"
     /// <param name="GraphConflictBehavior">Enum "Graph ConflictBehavior" value specifying the HttpHeader value</param>
     procedure SetMicrosftGraphConflictBehavior(GraphConflictBehavior: Enum "Graph ConflictBehavior")
     begin
-        SetQueryParameter('@microsoft.graph.conflictBehavior', Format(GraphConflictBehavior));
+        GraphOptionalParametersImpl.SetMicrosftGraphConflictBehavior(GraphConflictBehavior);
+    end;
+
+    internal procedure GetQueryParameters(): Dictionary of [Text, Text]
+    begin
+        exit(GraphOptionalParametersImpl.GetQueryParameters());
     end;
+    #endregion
 
+    #region ODataQueryParameters
 
-    local procedure SetQueryParameter(Header: Text; HeaderValue: Text)
+
+    /// <summary>
+    /// Sets the value for the OData Query Parameter '$count'.
+    /// see: https://learn.microsoft.com/en-us/graph/query-parameters?tabs=http#count-parameter
+    /// </summary>
+    procedure SetODataQueryParameterCount()
     begin
-        QueryParameters.Remove(Header);
-        QueryParameters.Add(Header, HeaderValue);
+        GraphOptionalParametersImpl.SetODataQueryParameterCount(true);
     end;
 
-    internal procedure GetQueryParameters(): Dictionary of [Text, Text]
+    /// <summary>
+    /// Sets the value for the OData Query Parameter '$expand'.
+    /// see: https://learn.microsoft.com/en-us/graph/query-parameters?tabs=http#expand-parameter
+    /// </summary>
+    /// <param name="Value">Text value specifying the query parameter</param>
+    procedure SetODataQueryParameterExpand("Value": Text)
     begin
-        exit(QueryParameters);
+        GraphOptionalParametersImpl.SetODataQueryParameterExpand("Value");
     end;
-    #endregion
 
-    var
-        QueryParameters: Dictionary of [Text, Text];
-        RequestHeaders: Dictionary of [Text, Text];
+    /// <summary>
+    /// Sets the value for the OData Query Parameter '$filter'.
+    /// see: https://learn.microsoft.com/en-us/graph/query-parameters?tabs=http#filter-parameter
+    /// </summary>
+    /// <param name="Value">Text value specifying the query parameter</param>
+    procedure SetODataQueryParameterFilter("Value": Text)
```

Comment:
Looking at all these new functions, I'm thinking it would be simpler to have an enum with Query Parameter options and then all these new functions simply turn into SetOdataQueryParameter(ODataQueryParameter: Enum "OData Query Parameter"; Value: Text). This way it's very easy to add another option, instead of having to add multiple functions here. It's also easier to see have an overview of what you can do in this codeunit. You can set query parameter, you can set request header etc. Then if you are interested in setting query parameter, you can see which are available in the enum. We may even be able to make this enum Extensible, which means Partners won't have to go add new functions every time they need a new ODataQueryParameter.

Reply:
TBH, I feel we should do the same for request header, simply have SetRequestHeader(RequestHeaderType: Enum "Request Header Type"; Value: Text);

Reply:
I wanted to make it as easy as possible by providing specific procedures for query parameters that only accept an integer value.
By making it an enum the partner would have to apply proper formatting to the value.

What I like on your suggestion is that it might be better for extensibility.

Maybe we can get a 2nd opinion on this topic before I start to refactor everrything.

Reply:
...which brings us back to the discussion if this isn't a fundamental building block, which should provide all capabilities out of the box, rather than being easily extensible. When it comes to system application modules, I pick "as easy as possible" over "extensibility" any day. But I can see the advantage of both solutions here.
@pri-kise, have you only implemented a small subset of query parameters so far? Are there many more to come?

Reply:
@JesperSchulz i added all the OData query parameters that are mentioned in the following list:
https://learn.microsoft.com/en-us/graph/query-parameters?tabs=http#odata-system-query-options

Reply:
@devgar since you provided a recent PR for this module what would you prefer?
@darjoo you've reviewed my original PR. What's your opinion?

Reply:
I feel I am no one to give my opinion, but I agree that the more flexibility is better, but this implementation seems complete enough for the MS Graph spec.

Reply:
As I see, it would be easy with an enum, you just have two Set functions SetRequestHeader() and SetODataQueryParameter(). As you code, you know whether you want a request header or a query parameter. Once you select it, you then get a list of only the valid headers or only the valid query parameters you can add.
Regarding the Integer, it's actually still stored as a text and provided to the web request as text itself.
So you can simply format it yourself, of course you can give it a string instead and mess it up, but you can do that with any other value as well if you provide an invalid value.

Reply:
I updated the code with new procedures and enums

Reply:
@AndreasMoth could you take another review after my changes? :)